delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/storages/operations/s3/client/S3StorageClient.scala

```diff
@@ -72,7 +88,75 @@ object S3StorageClient {
     override def headObject(bucket: String, key: String): IO[HeadObjectResponse] =
       client.headObject(HeadObjectRequest.builder().bucket(bucket).key(key).checksumMode(ChecksumMode.ENABLED).build)
 
-    override def underlyingClient: S3AsyncClientOp[IO] = client
+    override def objectExists(bucket: String, key: String): IO[Boolean] = {
+      headObject(bucket, key)
+        .redeemWith(
+          {
+            case _: NoSuchKeyException => IO.pure(false)
+            case e                     => IO.raiseError(e)
+          },
+          _ => IO.pure(true)
+        )
+    }
+
+    override def uploadFile(
+        fileData: Stream[IO, Byte],
+        bucket: String,
+        key: String,
+        algorithm: DigestAlgorithm
+    ): IO[UploadMetadata] = {
+      for {
+        fileSizeAcc <- Ref.of[IO, Long](0L)
+        digest      <- fileData
+                         .evalTap(_ => fileSizeAcc.update(_ + 1))
+                         .through(
+                           uploadFilePipe(bucket, key, algorithm)
+                         )
+                         .compile
+                         .onlyOrError
+        fileSize    <- fileSizeAcc.get
+        location     = baseEndpoint / bucket / Uri.Path(key)
+      } yield UploadMetadata(digest, fileSize, location)
+    }
+
+    private def uploadFilePipe(bucket: String, key: String, algorithm: DigestAlgorithm): Pipe[IO, Byte, String] = {
+      in =>
+        fs2.Stream.eval {
+          in.compile.to(Chunk).flatMap { chunks =>
+            val bs = chunks.toByteBuffer
+            for {
+              response <- client.putObject(
+                            PutObjectRequest
+                              .builder()
+                              .bucket(bucket)
+                              .deltaDigest(algorithm)
+                              .key(key)
+                              .build(),
+                            AsyncRequestBody.fromByteBuffer(bs)
+                          )
+            } yield {
+              checksumFromResponse(response, algorithm)
+            }
+          }
+        }
+    }
+
+    private def checksumFromResponse(response: PutObjectResponse, algorithm: DigestAlgorithm): String = {
+      algorithm.value match {
+        case "SHA-256" => Hex.encodeHexString(Base64.getDecoder.decode(response.checksumSHA256()))
+        case "SHA-1"   => Hex.encodeHexString(Base64.getDecoder.decode(response.checksumSHA1()))
+        case _         => throw new IllegalArgumentException(s"Unsupported algorithm for S3: ${algorithm.value}")
```

Comment:
I know we're not consistent but would ideally be in `IO` and can be deduplicated with below

Reply:
I would rather not because there's no I/O happening (also, the exception will never get hit due to config validation)